solidity/contracts/system/KeepFactorySelection.sol

```diff
@@ -191,11 +271,52 @@ library KeepFactorySelection {
             address(_self.factorySelector) == address(0),
             "Factory selector already set"
         );
+        require(address(_factorySelector) != address(0), "Invalid address");
+
+        _self.factorySelector = KeepFactorySelector(_factorySelector);
+    }
+
+    /// @notice Locks versions of factories. When lock is set vendor contracts
+    /// won't be called anymore to obtain the latest factories versions.
+    /// It requires expected factories addresses to be provided to protect from
+    /// locking on unexpected addresses.
+    /// @param _expectedKeepStakeFactory Expected KEEP-staked factory address
+    /// @param _expectedFullyBackedFactory Expected ETH-bond-only factory address
+    function lockFactoriesVersions(
+        Storage storage _self,
+        address _expectedKeepStakeFactory,
+        address _expectedFullyBackedFactory
+    ) internal {
+        require(!_self.factoriesVersionsLock, "Already locked");
+
+        require(
+            address(_self.keepStakeVendor) != address(0),
+            "KEEP backed vendor not set"
+        );
         require(
-            address(_factorySelector) != address(0),
-            "Invalid address"
+            address(_self.fullyBackedVendor) != address(0),
+            "Fully backed vendor not set"
         );
 
-        _self.factorySelector = KeepFactorySelector(_factorySelector);
+        address latestKeepStakeFactory = _self
+            .keepStakeVendor
+            .selectFactory();
+        address latestFullyBackedFactory = _self
+            .fullyBackedVendor
+            .selectFactory();
+
+        require(
+            address(latestKeepStakeFactory) == _expectedKeepStakeFactory,
+            "Unexpected KEEP backed factory"
+        );
+        require(
+            address(latestFullyBackedFactory) == _expectedFullyBackedFactory,
+            "Unexpected fully backed factory"
+        );
+
+        _self.keepStakeFactory = IBondedECDSAKeepFactory(latestKeepStakeFactory);
+        _self.fullyBackedFactory = IBondedECDSAKeepFactory(latestFullyBackedFactory);
+
+        _self.factoriesVersionsLock = true;
```

Comment:
Let's maybe lock before we call `selectFactory()` on any of the vendors. I don't expect any reentrancy issues here but better be safe than sorry.

Reply:
931ab0090a6329a62b1d378483d8045a031f3828